File: merge-sort.go
package main

func mergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	firstHalf := arr[:len(arr)/2]
	secondHalf := arr[len(arr)/2:]

	sortedFirstHalf := mergeSort(firstHalf)
	sortedSecondHalf := mergeSort(secondHalf)

	arr = []int{}

	i := 0
	j := 0

	for i < len(sortedFirstHalf) && j < len(sortedSecondHalf) {
		if sortedFirstHalf[i] <= sortedSecondHalf[j] {
			arr = append(arr, sortedFirstHalf[i])
			i++
		} else {
			arr = append(arr, sortedSecondHalf[j])
			j++
		}
	}

	for i < len(sortedFirstHalf) {
		arr = append(arr, sortedFirstHalf[i])
		i++
	}

	for j < len(sortedSecondHalf) {
		arr = append(arr, sortedSecondHalf[j])
		j++
	}

	return arr
}

File: merge-sort.py
def mergeSort(arr):
    if len(arr) < 2:
        return arr
    
    mid = len(arr) // 2
    firstHalf = arr[:mid]
    secondHalf = arr[mid:]

    sortedFirstHalf = mergeSort(firstHalf)
    sortedSecondHalf = mergeSort(secondHalf)

    sortedArr = []

    i = 0
    j = 0
    while i < len(sortedFirstHalf) and j < len(sortedSecondHalf):
        if sortedFirstHalf[i] <= sortedSecondHalf[j]:
            sortedArr.append(sortedFirstHalf[i])
            i += 1
        else:
            sortedArr.append(sortedSecondHalf[j])
            j += 1

    sortedArr.extend(sortedFirstHalf[i:])
    sortedArr.extend(sortedSecondHalf[j:])
    
    return sortedArr